### PR TITLE
BLE advertisement support 31 byte payload and added scanning

### PR DESCRIPTION
**Pull request by niklasad1**
Hi,
The major fixes with this PR are:
- Support payload size of 31 bytes instead of 30
- Support passive scanning of advertisements
- Added a sample application for scanning
- Added more inline comments

There are two controversial things in this PR which comes to mind:
- I'm using the radio HIL for callbacks to user-space
- I'm using the same buffer for both scanning and broadcasting (BLE_NONCONN_IND) because these never occur at the same time in current implementation!

Hopefully I managed to rebase properly this time.

--
Niklas A

### CODE REVIEW

File: chips/nrf51/src/ble_advertising_driver.rs [hunk 1]
```diff
@@ -196,29 +213,51 @@ impl<'a, A: hil::time::Alarm + 'a> BLE<'a, A> {
...
+                self.radio.start_adv_tx();
+            } else {
+                self.radio.start_adv_rx();
+            }
+            self.configure_periodic_alarm();
+        }
+    }
+}
+
+// FIXME: Temporary callback function *stolen* from the radio HIL
```
Comment (alevy) [inline chips/nrf51/src/ble_advertising_driver.rs]: @phil-levis _is_ the Radio trait meant to be generic beyond 802.15.4? (if not it's probably not the best name).

@niklasad1 assuming the answer is no, maybe just implement a trait interal to the `nrf51` crate?

Comment (niklasad1) [inline chips/nrf51/src/ble_advertising_driver.rs]: Sure, do you want the trait in a separate file or in [ble_advertising_driver.rs](https://github.com/helena-project/tock/blob/master/chips/nrf51/src/ble_advertising_driver.rs)?

Comment (alevy) [inline chips/nrf51/src/ble_advertising_driver.rs]: Same file is fine

**Comment by brghena** [comment]
@niklasad1 Is there anything holding this back still? I think the trait @alevy wanted is implemented (although the FIXME is still there).

It currently fails Travis, but once you fix that, if everything's good, we should probably merge this.

**Comment by niklasad1** [comment]
Sure, I'm back home again now I will check this tomorrow.
I think I made an ugly type-cast in the user-app that fails Travis. 
However I need to re-check the scanning one more time because I'm not completely sure that it works correctly.

**Pull request merged**